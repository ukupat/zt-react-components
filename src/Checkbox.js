import React, {PureComponent} from "react";
import PropTypes from "prop-types";
import styled from "styled-components";
import {borderRadiuses, colors, fontSizes} from "./designGuideline";
import Checkmark from "./icons/checkmark.svg";

const getBorderColor = ({checked, error}) => {
    if (error) {
        return colors.red;
    }
    else if (checked) {
        return colors.black;
    }

    return colors.gray;
};

const CheckboxElement = styled.span`
    display: inline-flex;
    align-items: center;
    justify-content: center;
    position: relative;

    margin-right: 7px;

    border: 1px solid ${getBorderColor};
    border-radius: ${borderRadiuses.default};

    background-color: ${({disabled}) => disabled && colors.darkerWhite};

    width: ${({size}) => size === "large" ? fontSizes.slightlyLarger : fontSizes.default};
    height: ${({size}) => size === "large" ? fontSizes.slightlyLarger : fontSizes.default};
    font-size: ${({size}) => size === "large" ? fontSizes.slightlyLarger : fontSizes.default};

    svg {
        opacity: ${({checked}) => checked ? 1 : 0}
    }

    &:hover:not([disabled]) {
        border-color: #000;
    }
`;

const Input = styled.input`
    position: absolute;
    top: 0;
    right: 0;
    bottom: 0;
    left: 0;
    opacity: 0;
`;

/**
 * Very sexy and simple checkbox (actually span) that can be:
 * - a bit larger,
 * - disabled,
 * - and of course checked.
 */
export default class Checkbox extends PureComponent {
    static propTypes = {
        size: PropTypes.oneOf(["small", "large"]),
        checked: PropTypes.bool,
        defaultChecked: PropTypes.bool,
        disabled: PropTypes.bool,
        error: PropTypes.bool,
        onChange: PropTypes.func,
        className: PropTypes.string,
    };

    constructor(props) {
        super(props);

        this.state = {
            isChecked: !!props.checked || !!props.defaultChecked
        };
    }

    componentWillReceiveProps(nextProps) {
        if (nextProps.checked !== this.props.checked) {
            this.setState({isChecked: nextProps.checked});
        }
    }

    onChange(evt) {
        const {onChange} = this.props;

        if (this.props.checked === undefined) {
            this.setState({isChecked: evt.target.checked});
        }

        onChange && onChange(evt);
    }

    render() {
        // picking out defaultChecked since we do not want to apply defaultChecked together with checked.
        // eslint-disable-next-line no-unused-vars
        const { size, disabled, defaultChecked, error, className, ...props } = this.props;
        const { isChecked } = this.state;

        return (
            <CheckboxElement
                className={className}
                checked={isChecked}
                error={error}
                size={size}
                disabled={disabled}
            >
                <Checkmark />

                <Input
                    {...props}
                    type="checkbox"
                    value={isChecked}
                    checked={isChecked}
                    onChange={this.onChange.bind(this)}
                    disabled={disabled}
                />
            </CheckboxElement>
        );
    }
}
