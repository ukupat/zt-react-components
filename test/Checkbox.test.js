import "jest-styled-components";
import React from "react";
import renderer from "react-test-renderer";
import ShallowRenderer from "react-test-renderer/shallow";
import Checkbox from "../src/Checkbox";

describe("<Checkbox />", function() {
    it("renders styled span element with input", function() {
        expect(renderer.create(<Checkbox name="test" />)).toMatchSnapshot();
    });

    it("renders checked", function() {
        expect(renderer.create(<Checkbox checked={true} />)).toMatchSnapshot();
    });

    it("renders error", function() {
        expect(renderer.create(<Checkbox error={true} />)).toMatchSnapshot();
    });

    it("renders disabled", function() {
        expect(renderer.create(<Checkbox disabled={true} />)).toMatchSnapshot();
    });

    it("renders larger", function() {
        expect(renderer.create(<Checkbox size="large" />)).toMatchSnapshot();
    });

    it("allows the styles to be overriden", function() {
        expect(renderer.create(<Checkbox className="custom-class" />)).toMatchSnapshot();
    });

    it("changes checked state if checked is change from the outside", function() {
        const renderer = new ShallowRenderer();
        renderer.render(<Checkbox checked={true} />);
        let result = renderer.getRenderOutput();
        expect(result.props.children[1].props.checked).toBe(true);

        renderer.render(<Checkbox checked={false} />);
        result = renderer.getRenderOutput();
        expect(result.props.children[1].props.checked).toBe(false);
    });
});
