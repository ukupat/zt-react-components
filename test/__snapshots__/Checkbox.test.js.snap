// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Checkbox /> allows the styles to be overriden 1`] = `
.iZWgNL {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  position: relative;
  margin-right: 7px;
  border: 1px solid #999;
  border-radius: 4px;
  background-color: ;
  width: 12px;
  height: 12px;
  font-size: 12px;
}

.iZWgNL svg {
  opacity: 0;
}

.iZWgNL:hover:not([disabled]) {
  border-color: #000;
}

.ewegQT {
  position: absolute;
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
  opacity: 0;
}

<span
  checked={false}
  className="custom-class sc-bdVaJa iZWgNL"
  disabled={undefined}
  size={undefined}
>
  <svg
    height="7"
    viewBox="0 0 9 7"
    width="9"
    xmlns="http://www.w3.org/2000/svg"
  >
    <title>
      P
    </title>
    <path
      d="M8.807.477A.625.625 0 0 0 8.35.283a.57.57 0 0 0-.44.194L3.217 5.17 1.09 3.043a.596.596 0 0 0-.44-.176.657.657 0 0 0-.457.176A.625.625 0 0 0 0 3.5c0 .176.064.328.193.457L2.76 6.523a.625.625 0 0 0 .457.194.57.57 0 0 0 .44-.194l5.15-5.15a.615.615 0 0 0 0-.896z"
      fill="#2F333D"
      fillRule="evenodd"
    />
  </svg>
  <input
    checked={false}
    className="sc-bwzfXH ewegQT"
    disabled={undefined}
    onChange={[Function]}
    type="checkbox"
    value={false}
  />
</span>
`;

exports[`<Checkbox /> renders checked 1`] = `
.hWREuk {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  position: relative;
  margin-right: 7px;
  border: 1px solid #000;
  border-radius: 4px;
  background-color: ;
  width: 12px;
  height: 12px;
  font-size: 12px;
}

.hWREuk svg {
  opacity: 1;
}

.hWREuk:hover:not([disabled]) {
  border-color: #000;
}

.ewegQT {
  position: absolute;
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
  opacity: 0;
}

<span
  checked={true}
  className="sc-bdVaJa hWREuk"
  disabled={undefined}
  size={undefined}
>
  <svg
    height="7"
    viewBox="0 0 9 7"
    width="9"
    xmlns="http://www.w3.org/2000/svg"
  >
    <title>
      P
    </title>
    <path
      d="M8.807.477A.625.625 0 0 0 8.35.283a.57.57 0 0 0-.44.194L3.217 5.17 1.09 3.043a.596.596 0 0 0-.44-.176.657.657 0 0 0-.457.176A.625.625 0 0 0 0 3.5c0 .176.064.328.193.457L2.76 6.523a.625.625 0 0 0 .457.194.57.57 0 0 0 .44-.194l5.15-5.15a.615.615 0 0 0 0-.896z"
      fill="#2F333D"
      fillRule="evenodd"
    />
  </svg>
  <input
    checked={true}
    className="sc-bwzfXH ewegQT"
    disabled={undefined}
    onChange={[Function]}
    type="checkbox"
    value={true}
  />
</span>
`;

exports[`<Checkbox /> renders disabled 1`] = `
.hNthpQ {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  position: relative;
  margin-right: 7px;
  border: 1px solid #999;
  border-radius: 4px;
  background-color: #f3f3f3;
  width: 12px;
  height: 12px;
  font-size: 12px;
}

.hNthpQ svg {
  opacity: 0;
}

.hNthpQ:hover:not([disabled]) {
  border-color: #000;
}

.ewegQT {
  position: absolute;
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
  opacity: 0;
}

<span
  checked={false}
  className="sc-bdVaJa hNthpQ"
  disabled={true}
  size={undefined}
>
  <svg
    height="7"
    viewBox="0 0 9 7"
    width="9"
    xmlns="http://www.w3.org/2000/svg"
  >
    <title>
      P
    </title>
    <path
      d="M8.807.477A.625.625 0 0 0 8.35.283a.57.57 0 0 0-.44.194L3.217 5.17 1.09 3.043a.596.596 0 0 0-.44-.176.657.657 0 0 0-.457.176A.625.625 0 0 0 0 3.5c0 .176.064.328.193.457L2.76 6.523a.625.625 0 0 0 .457.194.57.57 0 0 0 .44-.194l5.15-5.15a.615.615 0 0 0 0-.896z"
      fill="#2F333D"
      fillRule="evenodd"
    />
  </svg>
  <input
    checked={false}
    className="sc-bwzfXH ewegQT"
    disabled={true}
    onChange={[Function]}
    type="checkbox"
    value={false}
  />
</span>
`;

exports[`<Checkbox /> renders error 1`] = `
.jLfBzu {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  position: relative;
  margin-right: 7px;
  border: 1px solid #cd4936;
  border-radius: 4px;
  background-color: ;
  width: 12px;
  height: 12px;
  font-size: 12px;
}

.jLfBzu svg {
  opacity: 0;
}

.jLfBzu:hover:not([disabled]) {
  border-color: #000;
}

.ewegQT {
  position: absolute;
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
  opacity: 0;
}

<span
  checked={false}
  className="sc-bdVaJa jLfBzu"
  disabled={undefined}
  size={undefined}
>
  <svg
    height="7"
    viewBox="0 0 9 7"
    width="9"
    xmlns="http://www.w3.org/2000/svg"
  >
    <title>
      P
    </title>
    <path
      d="M8.807.477A.625.625 0 0 0 8.35.283a.57.57 0 0 0-.44.194L3.217 5.17 1.09 3.043a.596.596 0 0 0-.44-.176.657.657 0 0 0-.457.176A.625.625 0 0 0 0 3.5c0 .176.064.328.193.457L2.76 6.523a.625.625 0 0 0 .457.194.57.57 0 0 0 .44-.194l5.15-5.15a.615.615 0 0 0 0-.896z"
      fill="#2F333D"
      fillRule="evenodd"
    />
  </svg>
  <input
    checked={false}
    className="sc-bwzfXH ewegQT"
    disabled={undefined}
    onChange={[Function]}
    type="checkbox"
    value={false}
  />
</span>
`;

exports[`<Checkbox /> renders larger 1`] = `
.gDgQzI {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  position: relative;
  margin-right: 7px;
  border: 1px solid #999;
  border-radius: 4px;
  background-color: ;
  width: 14px;
  height: 14px;
  font-size: 14px;
}

.gDgQzI svg {
  opacity: 0;
}

.gDgQzI:hover:not([disabled]) {
  border-color: #000;
}

.ewegQT {
  position: absolute;
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
  opacity: 0;
}

<span
  checked={false}
  className="sc-bdVaJa gDgQzI"
  disabled={undefined}
  size="large"
>
  <svg
    height="7"
    viewBox="0 0 9 7"
    width="9"
    xmlns="http://www.w3.org/2000/svg"
  >
    <title>
      P
    </title>
    <path
      d="M8.807.477A.625.625 0 0 0 8.35.283a.57.57 0 0 0-.44.194L3.217 5.17 1.09 3.043a.596.596 0 0 0-.44-.176.657.657 0 0 0-.457.176A.625.625 0 0 0 0 3.5c0 .176.064.328.193.457L2.76 6.523a.625.625 0 0 0 .457.194.57.57 0 0 0 .44-.194l5.15-5.15a.615.615 0 0 0 0-.896z"
      fill="#2F333D"
      fillRule="evenodd"
    />
  </svg>
  <input
    checked={false}
    className="sc-bwzfXH ewegQT"
    disabled={undefined}
    onChange={[Function]}
    type="checkbox"
    value={false}
  />
</span>
`;

exports[`<Checkbox /> renders styled span element with input 1`] = `
.iZWgNL {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  position: relative;
  margin-right: 7px;
  border: 1px solid #999;
  border-radius: 4px;
  background-color: ;
  width: 12px;
  height: 12px;
  font-size: 12px;
}

.iZWgNL svg {
  opacity: 0;
}

.iZWgNL:hover:not([disabled]) {
  border-color: #000;
}

.ewegQT {
  position: absolute;
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
  opacity: 0;
}

<span
  checked={false}
  className="sc-bdVaJa iZWgNL"
  disabled={undefined}
  size={undefined}
>
  <svg
    height="7"
    viewBox="0 0 9 7"
    width="9"
    xmlns="http://www.w3.org/2000/svg"
  >
    <title>
      P
    </title>
    <path
      d="M8.807.477A.625.625 0 0 0 8.35.283a.57.57 0 0 0-.44.194L3.217 5.17 1.09 3.043a.596.596 0 0 0-.44-.176.657.657 0 0 0-.457.176A.625.625 0 0 0 0 3.5c0 .176.064.328.193.457L2.76 6.523a.625.625 0 0 0 .457.194.57.57 0 0 0 .44-.194l5.15-5.15a.615.615 0 0 0 0-.896z"
      fill="#2F333D"
      fillRule="evenodd"
    />
  </svg>
  <input
    checked={false}
    className="sc-bwzfXH ewegQT"
    disabled={undefined}
    name="test"
    onChange={[Function]}
    type="checkbox"
    value={false}
  />
</span>
`;
